src/nunit.analyzers/WithinUsage/WithinUsageAnalyzer.cs

```diff
@@ -0,0 +1,160 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using NUnit.Analyzers.Constants;
+using NUnit.Analyzers.Extensions;
+using NUnit.Analyzers.Helpers;
+
+namespace NUnit.Analyzers.WithinUsage
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class WithinUsageAnalyzer : BaseAssertionAnalyzer
+    {
+        private static readonly string[] SupportedIsMethods = new[]
+        {
+            NUnitFrameworkConstants.NameOfIsEqualTo,
+            NUnitFrameworkConstants.NameOfIsLessThan,
+            NUnitFrameworkConstants.NameOfIsLessThanOrEqualTo,
+            NUnitFrameworkConstants.NameOfIsGreaterThan,
+            NUnitFrameworkConstants.NameOfIsGreaterThanOrEqualTo,
+        };
+
+        private static readonly DiagnosticDescriptor descriptor = DiagnosticDescriptorCreator.Create(
+            id: AnalyzerIdentifiers.WithinIncompatibleTypes,
+            title: WithinUsageAnalyzerConstants.Title,
+            messageFormat: WithinUsageAnalyzerConstants.Message,
+            category: Categories.Assertion,
+            defaultSeverity: DiagnosticSeverity.Warning,
+            description: WithinUsageAnalyzerConstants.Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(descriptor);
+
+        protected override void AnalyzeAssertInvocation(OperationAnalysisContext context, IInvocationOperation assertOperation)
+        {
+            if (!AssertHelper.TryGetActualAndConstraintOperations(assertOperation, out _, out var constraintExpression))
+            {
+                return;
+            }
+
+            foreach (var constraintPart in constraintExpression.ConstraintParts)
+            {
+                if (!SupportedIsMethods.Contains(constraintPart.GetConstraintName()))
+                    continue;
+
+                // e.g. Is.EqualTo(expected).Within(0.1)
+                // Need to check type of expected
+
+                var withinSuffix = constraintPart.GetSuffix(NUnitFrameworkConstants.NameOfEqualConstraintWithin) as IInvocationOperation;
+
+                if (withinSuffix == null)
+                    continue;
+
+                var expectedType = constraintPart.GetExpectedArgument()?.Type;
+
+                if (expectedType == null || expectedType.TypeKind == TypeKind.Error)
+                    return;
```

Comment:
```suggestion
                if (withinSuffix is null)
                    continue;

                var expectedType = constraintPart.GetExpectedArgument()?.Type;

                if (expectedType is null || expectedType.TypeKind == TypeKind.Error)
                    return;
```

Reply:
I'm so relying on analyzers helping me to create consistent code, but here it failed me.

Even after enabling `dotnet_style_prefer_is_null_check_over_reference_equality_method = true:warning` I get none.
I hoped this would triggered [IDE0041](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0041) but it doesn't. It only triggers when using `ReferenceEquals(expectedType, null)`.
I see that the roslyn issue to [discourage ' == null'](https://github.com/dotnet/roslyn/issues/56859) is still open after two years. But the original reports also had created its own, so I added 'CSharpIsNullAnalyzer' to the project.